torchmultimodal/modules/layers/attention.py

```diff
@@ -0,0 +1,143 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+import itertools
+from typing import Optional
+
+import numpy as np
+
+import torch
+import torch.nn as nn
+
+from torchmultimodal.utils.common import tensor_slice
+
+
+class BroadcastPositionEmbedding(nn.Module):
+    r"""Spatiotemporal broadcasted positional embeddings.
+
+    Each embedding vector of the ``i``-th dim is repeated by ``N`` times, where
+    :math:`N = \prod_{j>i}\text{dim}[j]`.
+
+    Args:
+        shape (torch.Size): shape of raw data before batching and embedding
```

Comment:
let's do tuple for generality then since `torch.Size` is a [subclass of tuple](https://pytorch.org/docs/stable/generated/torch.Tensor.size.html#torch.Tensor.size) anyway.